C#: Set proxy environment variables, if Dependabot proxy is detected

mbg opened the pull request:
This PR is part of work to enable private package registries to be used in Default Setup. 

The existing Default Setup workflow will initialise the Dependabot package proxy, if a private package registry configuration is set. The host, port, and certificate used by the proxy are then passed to CodeQL in the `analyze` step. 

The changes in this PR modify the C# extractor to recognise when the corresponding environment variables are set. If so, we use the data from those environment variables to:

1. Construct the proxy address from the host and port and pass it to `dotnet` via the `HTTP_PROXY` and `HTTPS_PROXY` environment variables.
2. Store the certificate on disk, and pass the path to `dotnet` via `SSL_CERT_FILE`.
3. We also configure a suitable proxy for the feed reachability checks.

In testing so far, this works fine on Linux with fairly arbitrary versions of `dotnet`. It does not seem to work on macOS and likely also does not work on Windows.

michaelnebel commented on csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/NugetPackageRestorer.cs:
```diff
@@ -589,7 +593,26 @@ private static async Task ExecuteGetRequest(string address, HttpClient httpClien
+
+                if (!String.IsNullOrEmpty(this.dependabotProxy.CertificatePath))
+                {
+                    X509Certificate2 proxyCert = new X509Certificate2(this.dependabotProxy.CertificatePath);
```
This is because I haven't manage to fully manage to get the extractor to target .NET9 yet. I expect that when https://github.com/github/codeql/pull/18091 is merged then this can be fixed.

tamasvajk reviewed (commented):
LGTM, added some comments/questions.

What's not working on macOS? Is it the dotnet calls, or the httpclient that's not working?

mbg commented on csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependabotProxy.cs:
```diff
@@ -0,0 +1,77 @@
+        internal void ApplyProxy(ILogger logger, ProcessStartInfo startInfo)
+        {
+            // If the proxy isn't configured, we have nothing to do.
+            if (!this.IsConfigured) return;
```
Yeah, one way or another we'd have to check. I figured that putting it inside of the `DependabotProxy` implementation is simpler if we ever needed to use this elsewhere.

mbg commented on csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/NugetPackageRestorer.cs:
```diff
@@ -589,7 +593,26 @@ private static async Task ExecuteGetRequest(string address, HttpClient httpClien
+                httpClientHandler.Proxy = new WebProxy(this.dependabotProxy.Address);
+
+                if (!String.IsNullOrEmpty(this.dependabotProxy.CertificatePath))
+                {
```
Correct. The Dependabot Proxy uses a self-signed certificate that is generated when we initialise it in the Default Setup workflow. That's then passed to us in an environment variable, and we should trust it.

mbg commented:
> What's not working on macOS? Is it the dotnet calls

`dotnet` on macOS (and Windows) does not seem to respect `HTTP(S)_PROXY` and `SSL_CERT_FILE` (likely because it only uses OpenSSL under the hood on Linux?)

mbg commented on csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependabotProxy.cs:
```diff
@@ -0,0 +1,77 @@
+
+namespace Semmle.Extraction.CSharp.DependencyFetching
+{
+    public class DependabotProxy
```
Maybe, but I am not sure I see any immediate benefit to it. Although I guess it would simplify the implementation for the feed reachability checking? Happy to do that in a follow-up PR if it makes sense, but it would be nice to get this functionality shipped for now.

tamasvajk commented on csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependabotProxy.cs:
```diff
@@ -0,0 +1,92 @@
+            this.Certificate = new X509Certificate2(this.CertificatePath);
+        }
+
+        internal void ApplyProxy(ILogger logger, ProcessStartInfo startInfo)
```
`ApplyProxy` is specific to how this proxy should be applied to a `ProcessStartInfo`. Wouldn't it be clearer to move this to `DotNetCliInvoker`, and have something like the below? Then we would see in one place what environment variables are set on the subprocess for dotnet calls.
```C#
if (this.proxy != null)
{
  logger.LogInfo($"Setting up Dependabot proxy at {this.Address}");

  startInfo.EnvironmentVariables.Add("HTTP_PROXY", this.Address);
  startInfo.EnvironmentVariables.Add("HTTPS_PROXY", this.Address);
  startInfo.EnvironmentVariables.Add("SSL_CERT_FILE", this.certFile?.FullName);
}
```

And then you could probably even get rid to the `IsConfigured`.

tamasvajk commented on csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/NugetPackageRestorer.cs:
```diff
@@ -589,7 +593,25 @@ private static async Task ExecuteGetRequest(string address, HttpClient httpClien
+
+            // Configure the HttpClient to be aware of the Dependabot Proxy, if used.
+            HttpClientHandler httpClientHandler = new();
+            if (this.dependabotProxy.IsConfigured)
```
This `this.dependabotProxy.IsConfigured` could be replaced with a `this.dependabotProxy != null` check.

tamasvajk commented on csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependabotProxy.cs:
```diff
@@ -0,0 +1,92 @@
+            if (string.IsNullOrWhiteSpace(cert))
+            {
+                logger.LogInfo("No certificate configured for Dependabot proxy.");
+                return;
```
You could get rid of these early returns in the constructor if these checks were moved to a factory method that does the checks, and only creates a proxy instance if it is configured.

tamasvajk commented:
> `dotnet` on macOS (and Windows) does not seem to respect `HTTP(S)_PROXY` and `SSL_CERT_FILE` (likely because it only uses OpenSSL under the hood on Linux?)

If we don't expect this change to work on macos and windows, we should be explicit about it. Can we only create the dependabot proxy if we're on linux?

mbg commented:
> If we don't expect this change to work on macos and windows, we should be explicit about it. Can we only create the dependabot proxy if we're on linux?

@tamasvajk this is done as part of the Default Setup workflow with a dedicated action. Until recently, this only supported Linux. It does now support Windows and macOS as well, but the docs for the private registries feature will note the current limitations. In other words, it's not something we need to worry about in this C#-specific part.

tamasvajk commented:
> > If we don't expect this change to work on macos and windows, we should be explicit about it. Can we only create the dependabot proxy if we're on linux?
> 
> @tamasvajk this is done as part of the Default Setup workflow with a dedicated action. Until recently, this only supported Linux. It does now support Windows and macOS as well, but the docs for the private registries feature will note the current limitations. In other words, it's not something we need to worry about in this C#-specific part.

Does this mean that on Windows and MacOS, the `CODEQL_PROXY_*` environment variables are not going to be set?

mbg commented:
> Does this mean that on Windows and MacOS, the CODEQL_PROXY_* environment variables are not going to be set?

@tamasvajk They do get set on Windows and macOS. The environment variables we set here just wouldn't affect `dotnet` on those platforms (though it would be desirable if they did).

However, the proxy we install for the feed reachability check would be affected. That would result in a weird situation on those platforms where we check whether the feeds that require the proxy (i.e. require authentication) are reachable, determine that they are, but then `dotnet` cannot connect them.

So I have now added a check to skip initialising the `DependabotProxy` instance on those platforms.

I have also addressed all of your other review comments.

tamasvajk reviewed (approved):
LGTM

tamasvajk commented on csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/DependabotProxy.cs:
```diff
@@ -0,0 +1,87 @@
+
+                using var writer = certFile.CreateText();
+                writer.Write(cert);
+                writer.Close();
```
I'm somewhat surprised that this is needed. The `writer` is disposed at the end of this method, and we wouldn't access the file elsewhere before that. If you want to control the disposal a bit more fine grained, you could still use a `using` block, and you could get rid of the `Close` call.
```
using(var writer = certFile.CreateText()) 
{
  writer.Write(cert);
}
```
